### ISSUE TITLE
twisted mass clover mass rescale results in |source|=nan when more than one inversion is done

### ISSUE THREAD

**Issue by kostrzewa** [lattice/quda#453]
For multiple inversions with twisted mass + clover in the same job, the second inversion produces a `nan` for the source after mass rescaling. With the same calling code, multiple twisted mass (w/o clover) inversions go through without issues. This is with v0.8.0. I cannot test with the develop branch because `genericCopyColorSpinor` exits with `Not enabled` for the `QUDA_CHIRAL_GAMMA_BASIS` which is used by tmLQCD.

First inversion, point source at (0,0,0,0)[0,0] (XYZT)[sc]:

```
# QUDA: Using single prec. as sloppy!
# QUDA: Called _loadGaugeQuda
# QUDA: Using mixed precision CG!
# QUDA: mu = 0.006000, kappa = 0.137300
# QUDA: Using preconditioning!
# QUDA: time spent in reorder_spinor_toQuda: 0.147359 secs
# QUDA: Creating a DiracTwistedCloverPC operator (1 flavor(s))
# QUDA: Creating a DiracTwistedCloverPC operator (1 flavor(s))
# QUDA: Creating a DiracTwistedCloverPC operator (1 flavor(s))
# QUDA: Source: CPU = 1, CUDA copy = 1
# QUDA: Solution: CPU = 0, CUDA copy = 0
# QUDA: Prepared source = 0.854669
# QUDA: Prepared solution = 0
# QUDA: Prepared source post mass rescale = 0.854669
# QUDA: Creating a CG solver
# QUDA: CG: 0 iterations, <r,r> = 8.546690e-01, |r|/|b| = 1.000000e+00
[...]
# QUDA: CG: 3460 iterations, <r,r> = 8.516572e-19, |r|/|b| = 9.982365e-10
# QUDA: CG: Reliable updates = 5
# QUDA: CG: Convergence at 3460 iterations, L2 relative residual: iterated = 9.982365e-10, true = 9.982365e-10
# QUDA: Solution = 2.14837
# QUDA: Reconstructed: CUDA solution = 2.04531, CPU copy = 2.04531
# QUDA: Device memory used:  Spinor: 0.375000 GiB,  Gauge: 0.000000 GiB, Clover: 0.000000 GiB
# QUDA: Done: 3460 iter / 45.2157 secs = 798.678 Gflops
[...]
# Inversion done in 3460 iterations, squared residue = 7.325881e-19!
```

The next inversion (with a new source at (0,0,0,0)[0,1]) for some reason initiates auto-tuning and the prepared source has `nan` norm:

```
# QUDA: Using single prec. as sloppy!
# QUDA: Called _loadGaugeQuda
# QUDA: Using mixed precision CG!
# QUDA: mu = 0.006000, kappa = 0.137300
# QUDA: Using preconditioning!
# QUDA: Tuned block=(672,1,1), shared=16385 giving 0.00 Gflop/s, 119.77 GB/s for N4quda11CopyGaugeExIddLi18ENS_11FloatNOrderIdLi18ELi2ELi18EEES2_EE with out_stride=466560,in_stride=278528,geometery=4
# QUDA: Tuned block=(64,1,1), shared=24577 giving 0.00 Gflop/s, 112.78 GB/s for N4quda14ExtractGhostExIdLi18ELi4ELi0ENS_11FloatNOrderIdLi18ELi2ELi18EEEEE with prec=8,stride=466560,extract=1,dimension=0
# QUDA: Tuned block=(32,1,1), shared=24577 giving 0.00 Gflop/s, 67.60 GB/s for N4quda14ExtractGhostExIdLi18ELi4ELi0ENS_11FloatNOrderIdLi18ELi2ELi18EEEEE with prec=8,stride=466560,extract=0,dimension=0
# QUDA: Tuned block=(128,1,1), shared=12289 giving 0.00 Gflop/s, 301.32 GB/s for N4quda14ExtractGhostExIdLi18ELi4ELi1ENS_11FloatNOrderIdLi18ELi2ELi18EEEEE with prec=8,stride=466560,extract=1,dimension=1
# QUDA: Tuned block=(160,1,1), shared=24577 giving 0.00 Gflop/s, 294.81 GB/s for N4quda14ExtractGhostExIdLi18ELi4ELi1ENS_11FloatNOrderIdLi18ELi2ELi18EEEEE with prec=8,stride=466560,extract=0,dimension=1
# QUDA: Tuned block=(96,1,1), shared=6145 giving 0.00 Gflop/s, 323.95 GB/s for N4quda14ExtractGhostExIdLi18ELi4ELi2ENS_11FloatNOrderIdLi18ELi2ELi18EEEEE with prec=8,stride=466560,extract=1,dimension=2
# QUDA: Tuned block=(224,1,1), shared=24577 giving 0.00 Gflop/s, 321.99 GB/s for N4quda14ExtractGhostExIdLi18ELi4ELi2ENS_11FloatNOrderIdLi18ELi2ELi18EEEEE with prec=8,stride=466560,extract=0,dimension=2
# QUDA: Tuned block=(224,1,1), shared=24577 giving 0.00 Gflop/s, 336.53 GB/s for N4quda14ExtractGhostExIdLi18ELi4ELi3ENS_11FloatNOrderIdLi18ELi2ELi18EEEEE with prec=8,stride=466560,extract=1,dimension=3
# QUDA: Tuned block=(128,1,1), shared=12289 giving 0.00 Gflop/s, 338.31 GB/s for N4quda14ExtractGhostExIdLi18ELi4ELi3ENS_11FloatNOrderIdLi18ELi2ELi18EEEEE with prec=8,stride=466560,extract=0,dimension=3
# QUDA: Tuned block=(128,1,1)shared=0 giving -16.27 Gflop/s, 150.48 GB/s for N4quda12FmunuComputeIdNS_11FloatNOrderIdLi18ELi2ELi18EEES2_EE with threads=524288,stride=8,prec=0
# QUDA: Tuned block=(32,1,1), shared=0 giving 23.11 Gflop/s, 69.33 GB/s for N4quda13CloverComputeIdN11CloverOrder4quda11FloatNOrderIdLi72ELi2EEENS_11FloatNOrderIdLi18ELi2ELi18EEEEE with threads=524288,stride=262144,prec=8
# QUDA: Tuned block=(32,1,1), shared=0 giving 0.00 Gflop/s, 16.70 GB/s for N4quda12CloverInvertIdNS_11FloatNOrderIdLi72ELi2EEEEE with stride=262144,prec=8
# QUDA: Tuned block=(128,1,1), shared=0 giving 0.00 Gflop/s, 138.87 GB/s for N4quda10CopyCloverIfdLi72ENS_11FloatNOrderIfLi72ELi4EEENS1_IdLi72ELi2EEEEE with out_stride=262144,in_stride=262144
# QUDA: Tuned block=(64,1,1), shared=8193 giving 0.00 Gflop/s, 102.78 GB/s for N4quda10CopyCloverIsfLi72ENS_11FloatNOrderIsLi72ELi4EEENS1_IfLi72ELi4EEEEE with out_stride=262144,in_stride=262144
# QUDA: time spent in reorder_spinor_toQuda: 0.104045 secs
# QUDA: Creating a DiracTwistedCloverPC operator (1 flavor(s))
# QUDA: Creating a DiracTwistedCloverPC operator (1 flavor(s))
# QUDA: Creating a DiracTwistedCloverPC operator (1 flavor(s))
# QUDA: Source: CPU = 1, CUDA copy = 1
# QUDA: Solution: CPU = 0, CUDA copy = 0
# QUDA: Prepared source = -nan
# QUDA: Prepared solution = 0
# QUDA: Prepared source post mass rescale = -nan
# QUDA: Creating a CG solver
# QUDA: ERROR: Solver appears to have diverged (rank 3, host jrc0017, solver.cpp:134 in quda::Solver::PrintStats())
# QUDA:        last kernel called was (name=copyKernel,volume=16x32x32x16,aux=dst=vol=262144,stride=262144,precision=4,TwistFlavour=-1,src=vol=262144,stride=262144,precision=8,TwistFlavour=-1)
# QUDA: ERROR: Solver appears to have diverged (rank 1, host jrc0017, solver.cpp:134 in quda::Solver::PrintStats())
# QUDA:        last kernel called was (name=copyKernel,volume=16x32x32x16,aux=dst=vol=262144,stride=262144,precision=4,TwistFlavour=-1,src=vol=262144,stride=262144,precision=8,TwistFlavour=-1)
# QUDA: ERROR: Solver appears to have diverged (rank 2, host jrc0017, solver.cpp:134 in quda::Solver::PrintStats())
# QUDA:        last kernel called was (name=copyKernel,volume=16x32x32x16,aux=dst=vol=262144,stride=262144,precision=4,TwistFlavour=-1,src=vol=262144,stride=262144,precision=8,TwistFlavour=-1)
# QUDA: CG: 0 iterations, <r,r> = -nan, |r|/|b| = -nan
# QUDA: ERROR: Solver appears to have diverged (rank 0, host jrc0017, solver.cpp:134 in quda::Solver::PrintStats())
# QUDA:        last kernel called was (name=copyKernel,volume=16x32x32x16,aux=dst=vol=262144,stride=262144,precision=4,TwistFlavour=-1,src=vol=262144,stride=262144,precision=8,TwistFlavour=-1)
```

The tmLQCD quda interface can be found here: https://github.com/etmc/tmLQCD/blob/master/quda_interface.c

At program start, `_initQuda` is called which sets a number of parameters and then calls `initQuda`. 

For the inversion itself, `invert_eo_quda` is called. This translates the gauge field to the QUDA ordering and applies our twisted boundary conditions. Similarly, it translates the source spinor field, sets up the relevant parameters, calls `loadGaugeQuda` and `loadCloverQuda` and finally invertQuda. Afterwards it translates the solution back to our ordering.

Do you have an idea what could be going wrong here? 

I've already noticed that `freeCloverQuda()` is not called together with `freeGaugeQuda()`, but doing so does not alleviate the problem.

**Comment by AlexVaq** [comment]
Strange, for the disconnected we have always inverted 1000+ random sourced and never had a problem like this.

I also find strange that after calling invertQuda the tuned parameters are not saved.

Can you try disabling tuning? Just to see if there is something broken on that side, although I wouldn't discard a problem in the interface.

**Comment by kostrzewa** [comment]
> Can you try disabling tuning? Just to see if there is something broken on that side, although I wouldn't discard a problem in the interface.

This seems to be connected to auto-tuning. In addition, the number of iterations in the solve is reduced markedly (to around 2500) with auto-tuning disabled.

There is another issue which may or may not be related to this: for tmclover, the double-half solver is quite unstable and bails after

```
# QUDA: WARNING: CG: solver exiting due to too many true residual norm increases
```

**Comment by kostrzewa** [comment]
Please disregard the comment about the number of iterations, I was using NORMOP_SOLVE in this test rather than NORMERR_SOLVE.

**Comment by AlexVaq** [comment]
Half precision never worked very well with twisted-clover until dynamic-clover inversion was introduced. I don't understand very well why, to be frank. You can try enabling DYNAMIC_CLOVER to see if you get an improvement in stability, although mostl likely your solver will be slower.

**Comment by mathiaswagner** [comment]
Just trying to understand this better: As soon as you run without autotuning this works fine? When you turn on autotuning only the first inversion is fine and it crashes then?

Is a tunecache written to disk if you autotune?

If yes, does a subsequent run work after loading the tunecache (first iteration in this run)?

**Comment by kostrzewa** [comment]
> As soon as you run without autotuning this works fine? 

Yes.

> When you turn on autotuning only the first inversion is fine and it crashes then?

Yes, we check the residual externally and it is correct (to within an acceptable difference ```<r,r> 8.516572e-19(QUDA)  7.325881e-19(tmLQCD)```

To be precise, it doesn't crash, it just produces ```nan``` in the next inversion after mass rescaling and the solver then notices that it has "diverged". The norm of the input and solution spinors before mass rescaling are correct and agree between CPU and GPU. (see output above)

> Is a tunecache written to disk if you autotune?

Yes.

> If yes, does a subsequent run work after loading the tunecache (first iteration in this run)?

Yes. The first inversion loads the tunecache, goes through correctly (without redoing any tuning) and the subsequent inversion starts a tuning cycle and generates the ```nan```.

**Comment by mathiaswagner** [comment]
> Yes. The first inversion loads the tunecache, goes through correctly (without redoing any tuning) and the subsequent inversion starts a tuning cycle and generates the nan.

It looks very weird that the subsequent iteration retunes here. I assume that they all are called with the same parameters for tuning. I assume the inversion is called with the same setup and just uses a different source so all kernels should find tune parameters in the cache.
Figuring out what is going wrong here is the first thing we need to do.

**Comment by kostrzewa** [comment]
> I assume that they all are called with the same parameters for tuning. I assume the inversion is called with the same setup and just uses a different source so all kernels should find tune parameters in the cache.

Correct. It should be noted that the ```QudaInvertParam``` that we pass to ```invertQuda``` is created only once in our ```_initQuda``` and some parameters are set which we don't expect to change. At the next call of our Quda inverter interface, if the parameters changed (which they didn't in this case), these are adjusted and the the same struct is passed to ```invertQuda```. 

As a result, if ```invertQuda``` changes something in the parameter struct, these changes would persist to the next call of ```invertQuda```. Not also that this only happens with the clover term present, the normal TM operator works fine.

**Comment by maddyscientist** [comment]
Ok, I wonder if this reuse of the `QudaInvertParam` is significant.  Can you print it out before and after each call to `invertQuda`?
```c++
  /**
   * Print the members of QudaInvertParam.
   * @param param The QudaInvertParam whose elements we are to print.
   */
  void printQudaInvertParam(QudaInvertParam *param);
```

Also, can you instrument the `Dirac::prepare` method to print out the norms of the different vectors to see if we can isolate which function is getting the Nan?

**Comment by maddyscientist** [comment]
@kostrzewa 

I'm confused by this statement you made in the opening comment:

>  This is with v0.8.0. I cannot test with the develop branch because `genericCopyColorSpinor` exits with `Not enabled` for the `QUDA_CHIRAL_GAMMA_BASIS` which is used by tmLQCD.

If I look at the source for the copy routine, I see:
```c++
  /** Decide whether we are changing basis or not */
  template <typename FloatOut, typename FloatIn, int Ns, int Nc, typename OutOrder, typename InOrder>
    void genericCopyColorSpinor(OutOrder &outOrder, const InOrder &inOrder,
                                QudaGammaBasis dstBasis, QudaGammaBasis srcBasis,
                                const ColorSpinorField &out, QudaFieldLocation location) {
    if (dstBasis==srcBasis) {
      PreserveBasis<FloatOut, FloatIn, Ns, Nc> basis;
      PackSpinor<FloatOut, FloatIn, Ns, Nc, OutOrder, InOrder, PreserveBasis<FloatOut, FloatIn, Ns, Nc> >
	pack(outOrder, inOrder, basis, out, location);
      pack.apply(0);
    } else if (dstBasis == QUDA_UKQCD_GAMMA_BASIS && srcBasis == QUDA_DEGRAND_ROSSI_GAMMA_BASIS) {
      if (Ns != 4) errorQuda("Can only change basis with Nspin = 4, not Nspin = %d", Ns);
      if (Nc != 3) errorQuda("Can only change basis with Ncolor = 4, not Ncolor = %d", Nc);
      NonRelBasis<FloatOut, FloatIn, 4, 3> basis;
      PackSpinor<FloatOut, FloatIn, 4, 3, OutOrder, InOrder, NonRelBasis<FloatOut, FloatIn, 4, 3> >
        pack(outOrder, inOrder, basis, out, location);
      pack.apply(0);
    } else if (srcBasis == QUDA_UKQCD_GAMMA_BASIS && dstBasis == QUDA_DEGRAND_ROSSI_GAMMA_BASIS) {
      if (Ns != 4) errorQuda("Can only change basis with Nspin = 4, not Nspin = %d", Ns);
      if (Nc != 3) errorQuda("Can only change basis with Ncolor = 4, not Ncolor = %d", Nc);
      RelBasis<FloatOut, FloatIn, 4, 3> basis;
      PackSpinor<FloatOut, FloatIn, 4, 3, OutOrder, InOrder, RelBasis<FloatOut, FloatIn, 4, 3> >
        pack(outOrder, inOrder, basis, out, location);
      pack.apply(0);
    } else if (dstBasis == QUDA_UKQCD_GAMMA_BASIS && srcBasis == QUDA_CHIRAL_GAMMA_BASIS) {
      if (Ns != 4) errorQuda("Can only change basis with Nspin = 4, not Nspin = %d", Ns);
      if (Nc != 3) errorQuda("Can only change basis with Ncolor = 4, not Ncolor = %d", Nc);
      ChiralToNonRelBasis<FloatOut, FloatIn, 4, 3> basis;
      PackSpinor<FloatOut, FloatIn, 4, 3, OutOrder, InOrder, ChiralToNonRelBasis<FloatOut, FloatIn, 4, 3> >
	pack(outOrder, inOrder, basis, out, location);
      pack.apply(0);
      errorQuda("Not enabled");
    } else if (srcBasis == QUDA_UKQCD_GAMMA_BASIS && dstBasis == QUDA_CHIRAL_GAMMA_BASIS) {
      if (Ns != 4) errorQuda("Can only change basis with Nspin = 4, not Nspin = %d", Ns);
      if (Nc != 3) errorQuda("Can only change basis with Ncolor = 4, not Ncolor = %d", Nc);
      NonRelToChiralBasis<FloatOut, FloatIn, 4, 3> basis;
      PackSpinor<FloatOut, FloatIn, 4, 3, OutOrder, InOrder, NonRelToChiralBasis<FloatOut, FloatIn, 4, 3> >
        pack(outOrder, inOrder, basis, out, location);
      pack.apply(0);
    } else {
      errorQuda("Basis change not supported");
    }
  }
```
So it appears the the chiral basis should be supported for input and output into `QUDA_UKQCD_GAMMA_BASIS` which is what QUDA uses internally.  Can you give more details on the problem you've been seeing?

**Comment by kostrzewa** [comment]
@maddyscientist 

I'm guessing the relevant basis change is the following:

```
    } else if (dstBasis == QUDA_UKQCD_GAMMA_BASIS && srcBasis == QUDA_CHIRAL_GAMMA_BASIS) {
      if (Ns != 4) errorQuda("Can only change basis with Nspin = 4, not Nspin = %d", Ns);
      if (Nc != 3) errorQuda("Can only change basis with Ncolor = 4, not Ncolor = %d", Nc);
      ChiralToNonRelBasis<FloatOut, FloatIn, 4, 3> basis;
      PackSpinor<FloatOut, FloatIn, 4, 3, OutOrder, InOrder, ChiralToNonRelBasis<FloatOut, FloatIn, 4, 3> >
    pack(outOrder, inOrder, basis, out, location);
      pack.apply(0);
      errorQuda("Not enabled"); <-- CODE TERMINATES HERE
```

**Comment by maddyscientist** [comment]
Ah yes.  That's most likely a copy paste bug and if you delete that line it should fix it.

**Comment by maddyscientist** [comment]
@kostrzewa Can you check if this problem is still occurring with the current develop branch?  If it is, can you answer my question from https://github.com/lattice/quda/issues/453#issuecomment-210541239?  thx

**Comment by kostrzewa** [comment]
@maddyscientist I can only work on this very sporadically, I do have it on my radar and will get back to you.

**Comment by maddyscientist** [comment]
@kostrzewa No worries.  Just update me when you can.
